modules/electrophysiology_browser/php/models/datasettags.class.inc

```diff
@@ -0,0 +1,612 @@
+                    $hed_tags,
+                    function ($t) use ($tag) {
+                        return $tag['rel_id'] == $t['pair_rel_id'];
+                    }
+                );
+            }
+        );
+
+        $tag_names  = [];
+        $tag_string = '';
+
+        foreach ($root_tags as $root_tag) {
+            $tag_group    = [];
+            $group_member = $root_tag;
+
+            while ($group_member && !is_null($group_member['rel_id'])) {
+                $tag_group[]  = $group_member;
+                $member_pair  = array_values(
+                    array_filter(
+                        $hed_tags,
+                        function ($hed_tag) use ($group_member) {
+                            return !is_null($hed_tag['rel_id']) &&
+                            $hed_tag['rel_id'] == $group_member['pair_rel_id'];
+                        }
+                    )
+                );
+                $group_member = (count($member_pair) == 1)
+                ? $member_pair[0]
+                : null;
+            }
+
+            $subgroup_string = '';
+
+            foreach (array_reverse($tag_group) as $group_tag) {
+                if (is_null($group_tag['pair_rel_id'])) {
+                    $tag_string = $group_tag['name'] ?? 'n/a';
+                } else {
+                    if ($group_tag['has_pairing'] == 1) {
+                        if ($group_tag['additional_members'] > 0
+                            || strlen($subgroup_string) == 0
+                        ) {
+                            $comma_index     = $this
+                                    ->getNthTrailingMemberCommaIndex(
+                                        $tag_string,
+                                        $group_tag['additional_members'] + (
+                                            strlen($subgroup_string) > 0 ? 0 : 1
+                                        )
+                                    );
+                            $tag_string      = '(' .
+                              (is_null($group_tag['id'])
+                                  ? ''
+                                  : ($group_tag['name'] . ', ')
+                              ) .
+                              (strlen($subgroup_string) > 0
+                                  ? "$subgroup_string, "
+                                  : ''
+                              ) .
+                              substr($tag_string, 0, $comma_index) . ')' .
+                              substr($tag_string, $comma_index);
+                            $subgroup_string = '';
+                        } else {
+                            if (is_null($group_tag['id'])) {
+                                if (strlen($subgroup_string) > 0) {
+                                    $subgroup_string = "($subgroup_string)";
+                                } else {
+                                    error_log('"UNEXPECTED" STATE');  // TODO: raise
```

Comment:
This is not a useful error message to show up in the apache error log. No one reading it will know what it's talking about or even what module it's related to.

If it's necessary to log something here it should also be using a PSR logger so that the log level can be changed.

Reply:
Resolved by 91a08a0ebbf267b8934c7f797f7d81964be6947a.

Logging now uses PSR\Logger